### PR TITLE
Validate assocs length #2211

### PR DESCRIPTION
**Pull request by bartekupartek**

### CODE REVIEW

**Comment by josevalim** [comment]
Thanks @bartekupartek! I wonder if it would be better to set those on the cast_assoc/cast_embed functions instead of adding a new function to the API?

**Comment by bartekupartek** [comment]
I need it in `put_assoc` with `validate_length` cause I have to insert only existing tags, but I will check how it will behave in`cast_assoc`.

**Comment by josevalim** [comment]
> I need it in put_assoc with validate_length cause I have to insert only existing tags, but I will check how it will behave incast_assoc.

Hrm, that's a good point. I think what you have then is the way to go.

File: lib/ecto/changeset.ex [hunk 1]
```diff
@@ -1641,6 +1641,15 @@ defmodule Ecto.Changeset do
   defp codepoints_length(<<_, rest::binary>>, acc), do: codepoints_length(rest, acc + 1)
   defp codepoints_length(<<>>, acc), do: acc
 
+  defp list_length(%{types: types}, field, value) do
+    case Map.fetch(types, field) do
```
Comment (josevalim) [inline lib/ecto/changeset.ex]: This is going to break in a couple situations:

1. When the first element of the tuple is `:embed`
2. When the user has a custom type that works as a list

So I think we should pattern match rather on this:

```elixir
{:ok, {tag, relation}} when tag in [:embed, :assoc] ->
  ...
_ ->
  length(value)
```

File: lib/ecto/changeset/relation.ex [hunk 2]
```diff
@@ -39,6 +39,13 @@ defmodule Ecto.Changeset.Relation do
   def empty?(%{}, _), do: false
 
   @doc """
+  Filter empty changes
+  """
+  def filter_empty(%{cardinality: :many}, changes) do
```
Comment (josevalim) [inline lib/ecto/changeset/relation.ex]: Can you please make `empty?` use this function for the `cardinality: :many` case and then compare against an empty list?

**Comment by bartekupartek** [comment]
Ok thanks, I didn't considered `:embed` relation, I will try to correct this and `filter_empty`.

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**